File: packages/app-shell-contracts/src/runtime/task.ts
import type { ChatMessage, MessagePage } from "./chat.js";
import type { AgentCommandsCatalog, ConfigOptionsCatalog } from "./agent.js";
import type { IsolationKind, TaskStatus } from "./primitives.js";

export type TaskSummary = {
  task_id: string;
  project_id?: string;
  project_label?: string;
  title: string;
  title_source?: "prompt" | "agent" | "user";
  status: TaskStatus;
  task_version: number;
  message_history_version: number;
  has_messages: boolean;
  unread: boolean;
  pinned: boolean;
  attention?: TaskAttentionEvent;
  created_at: string;
  updated_at: string;
  last_activity: string;
  agent_id: string;
  agent_name: string;
  isolation: IsolationKind;
  workspace_root: string;
  worktree_id?: string;
  workspace_available?: boolean;
  worktree_name?: string;
  git_ref?: string;
};

export type TaskAttentionEvent = {
  event_id: string;
  reason: "finished" | "needsPermission" | "needsAnswer" | "stopped" | "failed";
  occurred_at: string;
};

export type TaskSnapshot = {
  lifecycle: "prepared" | "open" | "archived";
  task: TaskSummary;
  /** App Server-authored start of the active turn; absent when no turn is running. */
  active_turn_started_at?: string;
  chat: MessagePage;
  /** Active App Server requests render after durable Chat and never enter history. */
  active_requests: ChatMessage[];
  /** Durable follow-ups that have not entered Chat yet. */
  message_queue?: TaskMessageQueue;
  settings_summary: {
    agent_id: string;
    isolation: IsolationKind;
    model_id?: string;
  };
  agent_config?: ConfigOptionsCatalog;
  agent_commands?: AgentCommandsCatalog;
  preparation?:
    | { kind: "preparing" }
    | { kind: "ready" }
    | {
        kind: "blocked";
        blocker: {
          kind: "authRequired" | "setupRequired" | "nodeJsRequired" | "capabilityUnavailable" | "nativeSessionUnavailable";
        };
      }
    | { kind: "failed"; recovery?: "replaceTask" };
  send_capability: {
    state: "loading" | "ready" | "blocked" | "failed";
    blockers?: Array<{
      kind:
        | "taskPreparing"
        | "taskRunning"
        | "agentConfigNotReady"
        | "slashCommandsNotReady"
        | "attachmentsNeedRefresh"
        | "emptyMessage"
        | "missingRequiredOptions"
        | "failedValidation";
      message: string;
    }>;
  };
  input_capabilities?: {
    image: boolean;
  };
  context_usage?: TaskContextUsage;
  current_plan?: AgentPlan;
  revision: number;
  history_sync: HistorySyncState;
};

export type TaskMessageQueue = {
  revision: number;
  pause?: "restarted" | "unsuccessfulTurn" | "attachmentUnavailable";
  items: QueuedMessage[];
};

export type QueuedMessage = {
  queued_message_id: string;
  text: string;
  created_at: string;
  attachments?: Array<{ kind: string; label: string }>;
};

export type AgentPlan = {
  entries: AgentPlanEntry[];
};

export type AgentPlanEntry = {
  content: string;
  priority: "high" | "medium" | "low";
  status: "pending" | "in_progress" | "completed";
};

export type TaskContextUsage = {
  used_tokens: number;
  capacity_tokens: number;
  cost?: {
    amount: string;
    currency: string;
  };
  last_turn?: {
    total_tokens: number;
    input_tokens: number;
    output_tokens: number;
    reasoning_tokens?: number;
    cached_read_tokens?: number;
    cached_write_tokens?: number;
  };
};

export type HistorySyncState =
  | { state: "idle"; generation: number }
  | { state: "reloadAvailable"; generation: number }
  | { state: "syncing"; generation: number }
  | { state: "updated"; generation: number };

export type TaskListResult = {
  tasks: TaskSummary[];
  revision: number;
  lifecycle: "open" | "archived";
};

File: openaide-rs/app-server/src/protocol/model/task.rs
use crate::storage::records::{
    TaskAttentionEvent, TaskLifecycle, TaskNativeSessionDataFreshness, TaskPreparationRecord,
    TaskTitle,
};
use serde::{Deserialize, Serialize};

use super::{AgentCommandsCatalog, ConfigOptionCurrentValue, ConfigOptionsCatalog, MessagePage};

/// Current normalized ACP Agent Plan. Every incoming snapshot replaces this list in order.
#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct AgentPlan {
    pub entries: Vec<AgentPlanEntry>,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct AgentPlanEntry {
    pub content: String,
    pub priority: AgentPlanPriority,
    pub status: AgentPlanStatus,
}

#[derive(Debug, Clone, Copy, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentPlanPriority {
    High,
    Medium,
    Low,
}

#[derive(Debug, Clone, Copy, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentPlanStatus {
    Pending,
    InProgress,
    Completed,
}

#[derive(Debug, Clone, Copy, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum TaskStatus {
    Starting,
    Active,
    Stopping,
    Inactive,
    Failed,
    Completed,
    Waiting,
}

#[derive(Debug, Clone, Copy, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum IsolationKind {
    Local,
    GitWorktree,
    Docker,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct TaskSummary {
    pub task_id: String,
    pub title: Option<TaskTitle>,
    pub status: TaskStatus,
    pub task_version: u64,
    pub message_history_version: u64,
    pub unread: bool,
    pub pinned: bool,
    pub attention: Option<TaskAttentionEvent>,
    pub created_at: String,
    pub updated_at: String,
    pub last_activity: String,
    pub agent_id: String,
    pub agent_name: String,
    pub isolation: IsolationKind,
    pub workspace_root: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub project_root: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub worktree_id: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct TaskSnapshot {
    pub task: TaskSummary,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub active_turn_started_at: Option<String>,
    pub lifecycle: TaskLifecycle,
    pub chat: MessagePage,
    pub message_queue: crate::storage::records::TaskMessageQueueRecord,
    pub settings_summary: SettingsSummary,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub config_options_catalog: Option<ConfigOptionsCatalog>,
    pub native_session_data_freshness: TaskNativeSessionDataFreshness,
    /// Durable possible-external-change hint, projected as `historySync.reloadAvailable`.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub native_session_reload_requirement:
        Option<crate::storage::records::TaskNativeSessionReloadRequirement>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pending_config_change: Option<PendingTaskConfigChange>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub agent_commands_catalog: Option<AgentCommandsCatalog>,
    pub preparation: TaskPreparationRecord,
    pub supports_image_input: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context_usage: Option<TaskContextUsage>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub current_plan: Option<AgentPlan>,
    pub revision: u64,
}

/// Agent-reported usage for the live Native Session bound to this Task.
#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct TaskContextUsage {
    pub used_tokens: u64,
    pub capacity_tokens: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cost: Option<TaskUsageCost>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_turn: Option<TaskTurnUsage>,
}

/// Currency amount is stored as text so protocol state remains exact and Eq-safe.
#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct TaskUsageCost {
    pub amount: String,
    pub currency: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct TaskTurnUsage {
    pub total_tokens: u64,
    pub input_tokens: u64,
    pub output_tokens: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reasoning_tokens: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cached_read_tokens: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cached_write_tokens: Option<u64>,
}

/// Process-neutral projection of an in-flight config mutation.
/// Server sequencing remains private to the durable Task record.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct PendingTaskConfigChange {
    pub client_mutation_id: String,
    pub config_id: String,
    pub requested_value: ConfigOptionCurrentValue,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct SettingsSummary {
    pub agent_id: String,
    pub isolation: IsolationKind,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model_id: Option<String>,
}
